mod: erc-20

stephancill opened the pull request:
## Change Summary

This mod introduces a representation for ERC-20 tokens

- [x] Holder info
- [x] Price info
- [x] Token metadata
- [ ] Swap transaction

## Merge Checklist

<!-- Check the completed and unnecessary tasks below -->

- [ ] PR has a changeset
- [ ] PR includes documentation if necessary
- [ ] PR updates the [rich-embed examples](https://github.com/mod-protocol/mod/blob/main/examples/nextjs-shadcn/src/app/embeds.tsx) if necessary
- [ ] includes a parallel PR for [Mod-starter](https://github.com/mod-protocol/mod-starter) and the gateway if necessary

davidfurlong commented:
ready for review or no?

davidfurlong commented on examples/nextjs-shadcn/src/app/embeds.tsx:
```diff
@@ -71,6 +76,9 @@ export function Embeds(props: { embeds: Array<Embed> }) {
           address,
         },
+        farcaster: {
+          fid: "1214",
```
hm will this be a problem in prod? can we easily find connected users fid

stephancill replied:
I think it’s a fair assumption that clients should be able to provide the user’s FID, but we should explicitly handle the case where the FID isn’t available. Should only affect the holder info

if you’re referring to the example repo specifically then we can check if neynar can provide a reverse lookup for an fid given an address

davidfurlong commented:
People are going to love this!

Haven't been able to test fully, but I think there's unfortunately still quite a lot of work still to get this to be something that'll work well. 

1. Erc-20 endpoint is taking 8 seconds; what’s the slow request here
2. Skeleton loader instead of spinner
3. Token name shouldn’t be a button (its an external link), should be a link
4. Holders you know is really cool - but is this holders you follow? Should say that instead.
5. It’s unclear its an ERC-20 or even a token; we may need a design or adding the dollar sign to the name $DEGEN for example
6. It’s loading the /erc-20 endpoint 6 times for each page load right per mention - I understand dev mode doubles it but there’s still an issue here.
7. It’s being triggered by a CAIP-19 embed of an erc20, but nobody is sharing these right now, so it’ll never get triggered, and never get used.
8. Buy button press shouldn't replace the rest of the UI, should be added below & hide the buy button potentially. Buy button should be secondary not primary probably and stick to the right (similar to mint NFT action design)

Biggest issues here are 1) & 7) because they may not be fixable easily - what did you have in mind re 7)

davidfurlong commented:
the image should also link to the token page

stephancill commented:
@davidfurlong Thanks for the feedback!

1) Airstack is the slowest. Will adjust that to be loaded lazily. 

6. The endpoint is being hit 6 times initially because holder info, price info, and token metadata are separate requests with different cache controls executed in parallel. 3x requests are doubled with react strict mode in dev = 6 requests. In prod it wil be cached.

7. I’m expecting this to mostly be used in the context of a channel mod where the client will choose to manually associate it with a specific channel. I don’t see many links to tokens being shared, which is why I was thinking about cash tags. Let me know what your thoughts are on this. 

Noted the rest, will address soon 👍

davidfurlong commented:
> 1. Airstack is the slowest. Will adjust that to be loaded lazily.

Awesome that should help; they really responding in 6s in prod?

> 6. The endpoint is being hit 6 times initially because holder info, price info, and token metadata are separate requests with different cache controls executed in parallel. 3x requests are doubled with react strict mode in dev = 6 requests. In prod it wil be cached.

ah ok that makes sense, sounds good

> 7. I’m expecting this to mostly be used in the context of a channel mod where the client will choose to manually associate it with a specific channel. I don’t see many links to tokens being shared, which is why I was thinking about cash tags. Let me know what your thoughts are on this.

hm yeah currently it'd need a CAIP url shared as well as the channel mod or it'd have to be shown on each cast. Not ideal. Still a fan of cashtags triggering them, maybe there can be a cashtag registry or a registry book configured by the application

stephancill commented:
> Awesome that should help; they really responding in 6s in prod?

Their API only returns a limited number of rows at a time so we need to do multiple requests to get the balances of all the user that the user is following, and their cursor is sequential so we can't parallelize. So rn we just do the request and cache

davidfurlong commented:
ah probably the wrong API for this then. Or we could request an endpoint from airstack

qbig commented:
@stephancill @davidfurlong looking forward to testing this out

davidfurlong commented on examples/api/src/app/api/erc-20/route.ts:
```diff
@@ -0,0 +1,78 @@
+        }),
+      },
+    },
+  }[methodName];
```
this is a security issue

davidfurlong replied:
need to validate methodName is one of a certain set of valid strings first

davidfurlong commented on examples/api/src/app/api/erc-20/route.ts:
```diff
@@ -0,0 +1,78 @@
+  getTokenInfo,
+} from "./lib/utils";
+
+export async function GET(request: NextRequest) {
```
this whole function should be 3 separate endpoints - this abstraction is fairly complex to reason about

davidfurlong commented on examples/api/src/app/api/erc-20/route.ts:
```diff
@@ -0,0 +1,78 @@
+    });
+  }
+
+  const result = await fn({
```
this is returning HTML  - probably an error in HTML not json, which is causing the endpoint to 500

davidfurlong reviewed (changes requested):
It's still taking 6s frequently; we can't ship something that slow, even with a separate loader

Either we remove the parts that are causing this or we find another (quick) way to replace them. Might be worth pinging airstack

On the visual side, would be ideal if the margin bottom of the card wasn't bigger - seems a bit tricky to do unfortunately. It would also be ideal if the skeleton loader was close to the same width as the text when loaded

qbig commented:
is this about to get shipped?

davidfurlong commented:
> is this about to get shipped?

do you have some urgency to use this @qbig? feel free to DM me

stephancill commented:
Update:
- Implemented uniswap transactions with referral fee (#152 activates rev share). This takes noticeably longer to calculate swap routes than the 1inch API, so have introduced an intermediate "calculating best swap route" screen. There are caching options, but none that are serverless-compatible from what I can tell. 
- Refactored token info endpoints
- Investigated the pagination issue with Airstack - don't think they will be implementing in-query filtering any time soon
- Added an `href` property to the avatar element to also link to the token's page

I think the holders you follow feature is valuable despite the relatively long initial loading time (can be optimized in a future update)

Let me know what your thoughts are on merge readiness @davidfurlong

davidfurlong commented:
I'm getting 500s on localhost. There's something going wrong where in the example only one of the ERC20s loads

davidfurlong commented:
seems to almost be like the two requests for the two different erc20s compete with eachother

stephancill commented:
I think it's a rate limiting issue with the 1inch API that we're using for spot prices. Need to replace with uniswap SDK as well
